compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/utils.kt

```diff
@@ -533,3 +543,133 @@ private val PropertyDescriptor.declaresDefaultValue: Boolean
         is KtParameter -> declaration.defaultValue != null
         else -> false
     }
+
+@KspExperimental
+fun <T : Annotation> KSAnnotated.getAnnotationsByType(annotationKClass: KClass<T>): Sequence<T> {
+    return this.annotations.filter {
+        it.shortName.getShortName() == annotationKClass.simpleName && it.annotationType.resolve().declaration
+            .qualifiedName?.asString() == annotationKClass.qualifiedName
+    }.map { it.toAnnotation(annotationKClass.java) }
+}
+
+@KspExperimental
+fun <T : Annotation> KSAnnotated.isAnnotationPresent(annotationKClass: KClass<T>): Boolean =
+    getAnnotationsByType(annotationKClass).firstOrNull() != null
+
+@Suppress("UNCHECKED_CAST")
+private fun <T : Annotation> KSAnnotation.toAnnotation(annotationClass: Class<T>): T {
+    return Proxy.newProxyInstance(
+        annotationClass.classLoader,
+        arrayOf(annotationClass),
+        createInvocationHandler(annotationClass)
+    ) as T
+}
+
+@Suppress("TooGenericExceptionCaught")
+private fun KSAnnotation.createInvocationHandler(clazz: Class<*>): InvocationHandler {
+    val cache = ConcurrentHashMap<Pair<Class<*>, Any>, Any>(arguments.size)
```

Comment:
I was going to refactor this to use the `result` only as the key. However that is problematic in the cases with Byte or something that might cast to another result type. As you could have the same value for a Byte and an Integer return value but this would get overridden by the last cached value. So you might end up returning an Integer for a Byte type because the values were the same, which would cause a ClassCastException to occur. So keeping it as a `Pair`.